pinot-query-planner/src/main/java/org/apache/pinot/calcite/rel/rules/PinotJoinExchangeNodeInsertRule.java

```diff
@@ -102,7 +102,9 @@ private static PinotLogicalExchange createExchangeForLookupJoin(PinotHintOptions
       List<Integer> keys, RelNode child) {
     switch (distributionType) {
       case LOCAL:
-        return PinotLogicalExchange.create(child, RelDistributions.SINGLETON);
+        // NOTE: We use SINGLETON to represent local distribution. Add keys to the exchange because we might want to
+        //       switch it to HASH distribution to increase parallelism. See MailboxAssignmentVisitor for details.
```

Comment:
Reading RelDistribution.Type, shouldn't this be broadcast? The definition of broadcast is:

> There are multiple instances of the stream, and all records appear in each instance

While the definition of singleton is:

> There is only one instance of the stream. It sees all records.

Reply:
BTW, I don't get why we set DistributionType as SINGLETON in cases where we want to use HASH.

Reply:
This is not broadcast because we don't want to send data to other servers. This is not strictly SINGLETON if we want to add parallelism to local exchange (split one block into multiple and spread them into multiple operators). If there is no extra parallelism (1-to-1 distribution), then it is SINGLETON.